atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/ReloadingCloseableContainer.java

```diff
@@ -0,0 +1,222 @@
+ * 2. After a refresh, the new resource created must reflect the refreshed changes and not be closed.
+ * 3. After a refresh, there exists a time T when all resources except the latest resource is closed.
+ * 4. After {@link #close} has returned, the latest resource must be closed, and no further resources will be created.
+ *    (With 3, there exists a time T after {@link #close} has been called where all the resources are closed)
+ * 5. After {@link #close} has returned, all subsequent calls to runWith will fail.
+ *
+ * <b>When should you use this class?</b>
+ * You should use this when:
+ * 1. A resource is created from some live reloaded value, and new resources should be created when said value is
+ * updated.
+ * 2. The cost of retrying if the underlying resource is closed underneath you in an operation is expensive / not
+ * possible, OR there's risk of starvation (e.g with Cassandra, the topology changing in quick succession could
+ * starve existing connections if they're closed on every change)
+ * 3. It's acceptable for inflight consumers to use a stale resource (e.g at a temporary performance penalty for
+ * redirects)
+ *
+ */
+public final class ReloadingCloseableContainer<T extends Closeable> implements Closeable {
+    private static final SafeLogger log = SafeLoggerFactory.get(ReloadingCloseableContainer.class);
+
+    // This lock is marked fair to prevent starving #close. After close is called, no further #runWith
+    // calls will acquire the resource lock.
+    private final ReadWriteLock isClosedLock = new ReentrantReadWriteLock(true);
+    private final AtomicReference<Optional<Pair<ReadWriteLock, T>>> currentResourceWithLock =
+            new AtomicReference<>(Optional.empty());
+
+    private final Refreshable<Pair<ReadWriteLock, T>> refreshableResourceWithLock;
+    private final Disposable refreshableSubscriptionDisposable;
+
+    @GuardedBy("isClosedLock")
+    private volatile boolean isClosed;
+
+    public static <R, T extends Closeable> ReloadingCloseableContainer<T> of(
+            Function<R, T> resourceFactory, Refreshable<R> refreshableFactoryParameter) {
+        return new ReloadingCloseableContainer<>(resourceFactory, refreshableFactoryParameter);
+    }
+
+    private <R> ReloadingCloseableContainer(
+            Function<R, T> resourceFactory, Refreshable<R> refreshableFactoryParameter) {
+        isClosed = false;
+        refreshableResourceWithLock = refreshableFactoryParameter.map(factoryParameter -> runIfNotClosed(() -> {
+            T resource = resourceFactory.apply(factoryParameter);
+            // This lock does not need to be fair, since the isClosed lock is fair (preventing #close from being
+            // starved) and since subscribe happens after a map (and thus a new resource is created before attempting
+            // to close the old), there will not be any more readers of a resource about to be closed.
```

Comment:
This is untrue, as a thread could GC. What makes this relevant is that you only have one place in which you acquire a write lock, and only ever in a single thread (on closing) - if these were all write locks instead of mostly read locks, fairness would be required (but as it stands, an unfair lock remains correct, if only for reasons different to what you propose).